From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Connor Linfoot <connorlinfoot@me.com>
Date: Sun, 16 May 2021 15:07:34 +0100
Subject: [PATCH] Add basic Datapack API


diff --git a/src/main/java/io/papermc/paper/datapack/Datapack.java b/src/main/java/io/papermc/paper/datapack/Datapack.java
new file mode 100644
index 0000000000000000000000000000000000000000..7b2ab0be10a21e0496ad1d485ff8cb2c0b92a2cb
--- /dev/null
+++ b/src/main/java/io/papermc/paper/datapack/Datapack.java
@@ -0,0 +1,32 @@
+package io.papermc.paper.datapack;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+public interface Datapack {
+
+    /**
+     * @return the name of the pack
+     */
+    @NonNull
+    String getName();
+
+    /**
+     * @return the compatibility of the pack
+     */
+    @NonNull
+    Compatibility getCompatibility();
+
+    /**
+     * @return whether or not the pack is currently enabled
+     */
+    boolean isEnabled();
+
+    void setEnabled(boolean enabled);
+
+    enum Compatibility {
+        TOO_OLD,
+        TOO_NEW,
+        COMPATIBLE,
+    }
+
+}
diff --git a/src/main/java/io/papermc/paper/datapack/DatapackManager.java b/src/main/java/io/papermc/paper/datapack/DatapackManager.java
new file mode 100644
index 0000000000000000000000000000000000000000..58f78d5e91beacaf710f62461cf869f70d08b2a2
--- /dev/null
+++ b/src/main/java/io/papermc/paper/datapack/DatapackManager.java
@@ -0,0 +1,21 @@
+package io.papermc.paper.datapack;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+import java.util.Collection;
+
+public interface DatapackManager {
+
+    /**
+     * @return all the packs known to the server
+     */
+    @NonNull
+    Collection<Datapack> getPacks();
+
+    /**
+     * @return all the packs which are currently enabled
+     */
+    @NonNull
+    Collection<Datapack> getEnabledPacks();
+
+}
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 149c1a1cc39f79ca03e1f02cf2751885e0428895..30864584942415a42caa9d26de12a859456a0da2 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -327,9 +327,11 @@ public final class Bukkit {
     /**
      * Get the DataPack Manager.
      *
+     * @deprecated use {@link #getDatapackManager()}
      * @return the manager
      */
     @NotNull
+    @Deprecated(forRemoval = true, since = "1.20")
     public static DataPackManager getDataPackManager() {
         return server.getDataPackManager();
     }
@@ -2591,6 +2593,14 @@ public final class Bukkit {
     public static com.destroystokyo.paper.entity.ai.MobGoals getMobGoals() {
         return server.getMobGoals();
     }
+
+    /**
+     * @return the datapack manager
+     */
+    @NotNull
+    public static io.papermc.paper.datapack.DatapackManager getDatapackManager() {
+        return server.getDatapackManager();
+    }
     // Paper end
 
     @NotNull
diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index 9050d986b75482b5884a206f0980d1a23680001c..0d59a412ca47d922bb3386c7c16eae029c33fbdc 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -5565,6 +5565,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
      * @param world the world to check
      * @return true if this material can be used in this World.
      */
+    @Deprecated(forRemoval = true, since = "1.20") // Paper
     public boolean isEnabledByFeature(@NotNull World world) {
         if (isItem()) {
             return Bukkit.getDataPackManager().isEnabledByFeature(asItemType(), world);
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 0c71aa3b5758dbfdce8782e02772f9450901adbd..4a735cbe82eb02782452750823eca848613b6c8e 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -266,9 +266,11 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     /**
      * Get the DataPack Manager.
      *
+     * @deprecated use {@link #getDatapackManager()}
      * @return the manager
      */
     @NotNull
+    @Deprecated(forRemoval = true, since = "1.20") // Paper
     public DataPackManager getDataPackManager();
 
     /**
@@ -2260,5 +2262,11 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull
     com.destroystokyo.paper.entity.ai.MobGoals getMobGoals();
+
+    /**
+     * @return the datapack manager
+     */
+    @NotNull
+    io.papermc.paper.datapack.DatapackManager getDatapackManager();
     // Paper end
 }
diff --git a/src/main/java/org/bukkit/entity/EntityType.java b/src/main/java/org/bukkit/entity/EntityType.java
index b7a454ca11d484209d08d0003d4c19a431456687..0cdb4cbfddce1c138d7e2551c73713e616102123 100644
--- a/src/main/java/org/bukkit/entity/EntityType.java
+++ b/src/main/java/org/bukkit/entity/EntityType.java
@@ -461,6 +461,7 @@ public enum EntityType implements Keyed, Translatable, net.kyori.adventure.trans
      * @param world the world to check
      * @return true if this EntityType can be used to spawn an Entity for this World.
      */
+    @Deprecated(forRemoval = true, since = "1.20") // Paper
     public boolean isEnabledByFeature(@NotNull World world) {
         return Bukkit.getDataPackManager().isEnabledByFeature(this, world);
     }
diff --git a/src/main/java/org/bukkit/packs/DataPack.java b/src/main/java/org/bukkit/packs/DataPack.java
index 3d5af25a399589f1bdf95b77f584ae0ae5c26f02..0705e8a2b3281f17dace6a79f149a69672a42124 100644
--- a/src/main/java/org/bukkit/packs/DataPack.java
+++ b/src/main/java/org/bukkit/packs/DataPack.java
@@ -10,8 +10,10 @@ import org.jetbrains.annotations.NotNull;
  * Represents a data pack.
  *
  * @see <a href="https://minecraft.wiki/w/Data_pack">Minecraft wiki</a>
+ * @deprecated use {@link io.papermc.paper.datapack.Datapack}
  */
 @ApiStatus.Experimental
+@Deprecated(forRemoval = true, since = "1.20") // Paper
 public interface DataPack extends Keyed {
 
     /**
diff --git a/src/main/java/org/bukkit/packs/DataPackManager.java b/src/main/java/org/bukkit/packs/DataPackManager.java
index c33ca7b86426223200efa7df53faef061c3c7c0b..8e6e874c997985d0b02d572775b5f59065ff2cd3 100644
--- a/src/main/java/org/bukkit/packs/DataPackManager.java
+++ b/src/main/java/org/bukkit/packs/DataPackManager.java
@@ -13,8 +13,10 @@ import org.jetbrains.annotations.Nullable;
 
 /**
  * Manager of data packs.
+ * @deprecated use {@link io.papermc.paper.datapack.DatapackManager}
  */
 @ApiStatus.Experimental
+@Deprecated(forRemoval = true, since = "1.20") // Paper
 public interface DataPackManager {
 
     /**
