From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jakub Zacek <dawon@dawon.eu>
Date: Mon, 17 Jun 2024 19:07:00 +0200
Subject: [PATCH] Fixes and additions to the SpawnReason API


diff --git a/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java b/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
index 394ea92c1021d04e1687b8c12854356265e3d5de..a8423c12dfc088bfa5e2696dd0ef954c91a742d7 100644
--- a/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
+++ b/src/main/java/org/bukkit/event/entity/CreatureSpawnEvent.java
@@ -202,6 +202,12 @@ public class CreatureSpawnEvent extends EntitySpawnEvent {
          * When a creature is spawned by an enchantment
          */
         ENCHANTMENT,
+        // Paper start - Fixes and additions to the SpawnReason API
+        /**
+         * When an entity spawns from an ominous item spawner
+         */
+        OMINOUS_ITEM_SPAWNER,
+        // Paper end - Fixes and additions to the SpawnReason API
         /**
          * When a creature is spawned by plugins
          */
