From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 28 Mar 2016 20:32:58 -0400
Subject: [PATCH] Entity AddTo/RemoveFrom World Events


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 5975cc2fa72609ea5f3e6f99155d6e5bc321a321..38d27d84c985724d1f58bbb7c8f19842cf397074 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -2161,6 +2161,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
                 entity.setOrigin(entity.getOriginVector().toLocation(getWorld()));
             }
             // Paper end - Entity origin API
+            new com.destroystokyo.paper.event.entity.EntityAddToWorldEvent(entity.getBukkitEntity(), ServerLevel.this.getWorld()).callEvent(); // Paper - fire while valid
         }
 
         public void onTrackingEnd(Entity entity) {
@@ -2236,6 +2237,13 @@ public class ServerLevel extends Level implements WorldGenLevel {
                 }
             }
             // CraftBukkit end
+            // Paper start - fire while valid
+            com.destroystokyo.paper.event.entity.EntityRemoveFromWorldEvent.RemovalReason removalReason = com.destroystokyo.paper.event.entity.EntityRemoveFromWorldEvent.RemovalReason.UNKNOWN;
+            Entity.RemovalReason reason = entity.getRemovalReason();
+            if (reason != null) removalReason = com.destroystokyo.paper.event.entity.EntityRemoveFromWorldEvent.RemovalReason.values()[reason.ordinal()];
+            else if (entity.chunkStatus == FullChunkStatus.INACCESSIBLE) removalReason = com.destroystokyo.paper.event.entity.EntityRemoveFromWorldEvent.RemovalReason.UNLOADED_TO_CHUNK;
+            new com.destroystokyo.paper.event.entity.EntityRemoveFromWorldEvent(entity.getBukkitEntity(), ServerLevel.this.getWorld(), removalReason).callEvent();
+            // Paper end
         }
 
         public void onSectionChange(Entity entity) {
