From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date: Sun, 3 Mar 2024 18:05:10 +0100
Subject: [PATCH] server gen test


diff --git a/build.gradle.kts b/build.gradle.kts
index 56106696ffb9aefb0a17c9b0e434bca52ea23634..129e42e8a05e13c2fbdabf4011c5a6db970ab559 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -182,17 +182,31 @@ fun TaskContainer.registerRunTask(
 }
 
 // Paper start
+val generatedServerDir: File = rootProject.file("paper-generator/generatedServer")
+
 sourceSets.main {
     java {
+        srcDir(generatedServerDir)
         exclude("excluded/**")
     }
 }
 
 idea {
     module {
+        generatedSourceDirs.add(generatedServerDir)
         excludeDirs.add(projectDir.resolve("src/main/java/excluded"))
     }
 }
+
+val scanJarForOldGeneratedCode = tasks.register("scanJarForOldGeneratedCode", io.papermc.paperweight.tasks.ScanJarForOldGeneratedCode::class) {
+    mcVersion.set(providers.gradleProperty("mcVersion"))
+    annotation.set("Lio/papermc/paper/generated/GeneratedFrom;")
+    jarToScan.set(tasks.jar.flatMap { it.archiveFile })
+    classpath.from(configurations.compileClasspath)
+}
+tasks.check {
+    dependsOn(scanJarForOldGeneratedCode)
+}
 // Paper end
 
 val runtimeClasspathWithoutVanillaServer = configurations.runtimeClasspath.flatMap { it.elements }
